README.md
# Unsupervised Learning and Segmentation of Complex Activities from Video

Python implementation of the paper Unsupervised Learning and Segmentation of Complex Activities from Video. [a link](https://arxiv.org/pdf/1803.09490.pdf)

## Results and changes:
Full log for the Breakfast dataset result is here.  [a link](https://github.com/Annusha/slim_mallow/blob/master/results/log)  
Reduced log just with total accuracy and intermidiate one is here.  [a link](https://github.com/Annusha/slim_mallow/blob/master/results/evaluation.md)  
Changes in the model are here.  [a link](https://github.com/Annusha/slim_mallow/blob/master/results/changes.md)  

## Preceding steps:

### To create environment:

```conda create --name <env> --file requirements.txt```

### Data:

At this moment it's easy to run for the Breakfast dataset, for any other there should be made some changes (description for it will be added if necessary).

#### Breakfast Dataset
For each frame there should be precomputed feature vector. (originally dense trajetories).  
For each video there should be represented as one file contained matrix (row is a frame feature) which can be loaded via np.load (e.g. txt file, gz archive).   

--dataset_root=/path/to/your/folder : root for everything, there are all other subfolders located.  
--data=whatever_you_want  
then feature location is /dataset_root/feat/whatever_you_want/ascii  
All features of the same precomputed type (or folders with them, any hierarchy) should be eventually in /dataset_root/data/ascii/here_your_features  

--gt=folder_name: in folder /dataset_root/data should locate --gt=folder_name (default name groundTruth, full path: /dataset_root/data/gt) which contains ground truth labels for each your training file with the same name but without extension (txt/gz).  
If you already have mapping.txt file (correspondences between numerical indexes and subactivity names, look at data_ex/mapping.txt) put this file into /dataset_root/data/gt/mapping/mapping.txt.  
If you don't have such file and gt segmentation is in the format as for YTI (data_ex/YTI_seg_ex) you can create it from the ground truth segmentation:   
```python YTI_utils/mapping.py```  
Comment 27,28 lines if you do not need background class, gt_folder (line 21) - your folder with ground truth labels  

--feature_dim=64 : dimensionality of frame feature  
--end=txt : extension of your stored features (txt/gz/whatever)  
--data_type=4 (default) : others types are different precomputed features for breakfast.  

### Other important parameters:

--gmms=many : paper version of Gaussian Mixture Models for subactions  
--gmms=one : slim version (more than in 4 time faster)  

--ordering=True: using Generalized Mallow Model to create order (default=True)  

--full=False : quick check on the part of the videos (take first 20 videos, default=True, namely all data will be readed)  
  
--save_model=False : to save trained embedding weights  

--prefix=whatever : prefix for you data names which will be stored, namely log file, segmentation for each video (can't turn off), embedding weights (if specified).  

### How to run:

```
python pipeline.py --your_params
```

### Full set of parameters

```
usage: pipeline.py [-h] [--subaction SUBACTION] [--dataset DATASET]
                   [--data_type DATA_TYPE] [--dataset_root DATASET_ROOT]
                   [--data DATA] [--gt GT] [--feature_dim FEATURE_DIM]
                   [--ext EXT] [--seed SEED] [--lr LR] [--lr_adj LR_ADJ]
                   [--momentum MOMENTUM] [--weight_decay WEIGHT_DECAY]
                   [--batch_size BATCH_SIZE] [--num_workers NUM_WORKERS]
                   [--embed_dim EMBED_DIM] [--epochs EPOCHS]
                   [--gt_training GT_TRAINING] [--gmm GMM] [--gmms GMMS]
                   [--reg_cov REG_COV] [--ordering ORDERING] [--bg BG]
                   [--bg_trh BG_TRH] [--viterbi VITERBI]
                   [--save_model SAVE_MODEL] [--resume RESUME]
                   [--resume_str RESUME_STR]
                   [--save_likelihood SAVE_LIKELIHOOD]
                   [--save_embed_feat SAVE_EMBED_FEAT] [--full FULL]
                   [--zeros ZEROS] [--vis VIS] [--prefix PREFIX] [--log LOG]
                   [--log_str LOG_STR]

optional arguments:
  -h, --help            show this help message and exit
  --subaction SUBACTION
                        measure accuracy for different subactivities
  --dataset DATASET     Breakfast dataset (bf) or YouTube Instructional (yti)
  --data_type DATA_TYPE
                        for this moment valid just for Breakfast dataset0:
                        kinetics - features from the stream network1: data -
                        normalized features2: s1 - features without
                        normalization3: videos4: new features, not specified
                        earlier
  --dataset_root DATASET_ROOT
                        root folder for dataset:Breakfast / YTInstructions
  --data DATA           direct path to your data features
  --gt GT               folder with ground truth labels
  --feature_dim FEATURE_DIM
                        feature dimensionality
  --ext EXT             extension of the feature files
  --seed SEED           seed for random algorithms, everywhere
  --lr LR               initial learning rate
  --lr_adj LR_ADJ       will lr be multiplied by 0.1 in the middle
  --momentum MOMENTUM   momentum
  --weight_decay WEIGHT_DECAY
                        regularization constant for l_2 regularizer of W
  --batch_size BATCH_SIZE
                        batch size for training embedding (default: 40)
  --num_workers NUM_WORKERS
                        number of threads for dataloading
  --embed_dim EMBED_DIM
                        number of dimensions in embedded space
  --epochs EPOCHS       number of epochs for training embedding
  --gt_training GT_TRAINING
                        training embedding (rank model) either with gt labels
                        or with labels gotten from the temporal model
  --gmm GMM             number of components for gaussians
  --gmms GMMS           number of gmm for the video collection: many/one
  --reg_cov REG_COV     gaussian mixture model parameter
  --ordering ORDERING   apply Mallow model to incorporate ordering
  --bg BG               if we need to apply part for modeling background
  --bg_trh BG_TRH
  --viterbi VITERBI
  --save_model SAVE_MODEL
                        save embedding model after training
  --resume RESUME       load model for embeddings, if positive then it is
                        number of epoch which should be loaded
  --resume_str RESUME_STR
                        which model to load
  --save_likelihood SAVE_LIKELIHOOD
  --save_embed_feat SAVE_EMBED_FEAT
                        save features after embedding trained on gt
  --full FULL           check smth using only 15 videos
  --zeros ZEROS         if True there can be SIL label (beginning and end)that
                        is zeros labels relating to non actionif False then 0
                        labels are erased from ground truth labeling at all
                        (SIL label), only for Breakfast dataset
  --vis VIS             save visualisation of embeddings
  --prefix PREFIX       prefix for log file
  --log LOG             DEBUG | INFO | WARNING | ERROR | CRITICAL
  --log_str LOG_STR     unify all savings

```




results/changes.md
### Speeding up

As far as we made an assumption of frames interindependence, video likelihood P(F_m | l_m) can be splitted into product of frame likelihoods given current label assignment. 

![alt text](https://github.com/Annusha/slim_mallow/blob/master/results/1.png)

Authors of the original paper modeled likelihood for each video F_m based on the entire video collection, but excluded the very same video.  Due to the definition of Gibbs sampling procedure, the probability of the video to have respective sub-activity frame-wise assignments is conditioned on the segmentation of the rest of the video collection. 
To obtain the exact values of likelihood for every frame, the Gaussian Mixture Model is fitted for each video separately.

![alt text](https://github.com/Annusha/slim_mallow/blob/master/results/2.png)

where Q is empirically defined number of components. 
Whereas we train the embedding to form distinct sub-activity clusters given the initial resp. updated segmentation at the beginning of each iteration, the exclusion of all frames belonging to the video provided that the accumulation of the rest frames does not affect the statistics of fitted gaussians noticeably. For the simplicity and quick computations we streamline the model and keep just one Gaussian Mixture Model per sub-action instead of N_m mixtures that give us

![alt text](https://github.com/Annusha/slim_mallow/blob/master/results/3.png)



utils/arg_pars.py
#!/usr/bin/env python

"""Hyper-parameters and logging set up

opt: include all hyper-parameters
logger: unified logger for the project
"""

__all__ = ['opt']
__author__ = 'Anna Kukleva'
__date__ = 'August 2018'

import argparse


parser = argparse.ArgumentParser()

###########################################
# data
parser.add_argument('--subaction', default='all',  # ['changing_tire', 'coffee', 'jump_car', 'cpr', 'repot']
                    help='measure accuracy for different subactivities')
parser.add_argument('--dataset', default='bf',
                    help='Breakfast dataset (bf) or YouTube Instructional (yti)')
parser.add_argument('--data_type', default=4, type=int,
                    help='for this moment valid just for Breakfast dataset'
                         '0: kinetics - features from the stream network'
                         '1: data - normalized features'
                         '2: s1 - features without normalization'
                         '3: videos'
                         '4: new features, not specified earlier')


parser.add_argument('--dataset_root', default='/media/data/kukleva/lab/Breakfast',
                    help='root folder for dataset:'
                         'Breakfast / YTInstructions')
parser.add_argument('--data', default='feat',
                    help='direct path to your data features')
parser.add_argument('--gt', default='groundTruth',
                    help='folder with ground truth labels')
parser.add_argument('--feature_dim', default=64,
                    help='feature dimensionality')
parser.add_argument('--ext', default='gz',
                    help='extension of the feature files')


###########################################
# hyperparams parameters for embeddings
parser.add_argument('--seed', default=0,
                    help='seed for random algorithms, everywhere')
parser.add_argument('--lr', default=1e-10, type=float,
                    help='initial learning rate')
parser.add_argument('--lr_adj', default=False, type=bool,
                    help='will lr be multiplied by 0.1 in the middle')
parser.add_argument('--momentum', default=0.9,
                    help='momentum')
parser.add_argument('--weight_decay', default=1e-4,
                    help='regularization constant for l_2 regularizer of W')
parser.add_argument('--batch_size', default=256,
                    help='batch size for training embedding (default: 40)')
parser.add_argument('--num_workers', default=4,
                    help='number of threads for dataloading')
parser.add_argument('--embed_dim', default=30, type=int,
                    help='number of dimensions in embedded space')
parser.add_argument('--epochs', default=12, type=int,
                    help='number of epochs for training embedding')
parser.add_argument('--gt_training', default=False, type=bool,
                    help='training embedding (rank model) either with gt labels '
                         'or with labels gotten from the temporal model')


###########################################
# probabilistic parameters
parser.add_argument('--gmm', default=1, type=int,
                    help='number of components for gaussians')
parser.add_argument('--gmms', default='one',
                    help='number of gmm for the video collection: many/one')
parser.add_argument('--reg_cov', default=1e-1, type=float,
                    help='gaussian mixture model parameter')
parser.add_argument('--ordering', default=True,
                    help='apply Mallow model to incorporate ordering')


###########################################
# bg
parser.add_argument('--bg', default=False, type=bool,
                    help='if we need to apply part for modeling background')
parser.add_argument('--bg_trh', default=85, type=int)


###########################################
# viterbi
parser.add_argument('--viterbi', default=False, type=bool)

###########################################
# save
parser.add_argument('--save_model', default=False, type=bool,
                    help='save embedding model after training')
parser.add_argument('--resume', default=False, type=bool,
                    help='load model for embeddings, if positive then it is number of '
                         'epoch which should be loaded')
parser.add_argument('--resume_str',
                    default='',
                    help='which model to load')
parser.add_argument('--save_likelihood', default=False, type=bool)
parser.add_argument('--save_embed_feat', default=False,
                    help='save features after embedding trained on gt')

###########################################
# additional
parser.add_argument('--full', default=True, type=bool,
                    help='check smth using only 15 videos')
parser.add_argument('--zeros', default=False, type=bool,
                    help='if True there can be SIL label (beginning and end)'
                         'that is zeros labels relating to non action'
                         'if False then 0 labels are erased from ground truth '
                         'labeling at all (SIL label), only for Breakfast dataset')
parser.add_argument('--vis', default=False, type=bool,
                    help='save visualisation of embeddings')
parser.add_argument('--prefix', default='slim.mallow.',
                    help='prefix for log file')

###########################################
# logs
parser.add_argument('--log', default='DEBUG',
                    help='DEBUG | INFO | WARNING | ERROR | CRITICAL')
parser.add_argument('--log_str', default='',
                    help='unify all savings')

opt = parser.parse_args()



